cgen error with nested match exprs

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Linux version 5.14.12-arch1-1 (linux@archlinux) (gcc (GCC) 11.1.0, GNU ld (GNU Binutils) 2.36.1) #1 SMP PREEMPT Wed, 13 Oct 2021 16:58:16 +0000
Processor: 12 cpus, 64bit, little endian, AMD Ryzen 5 2600X Six-Core Processor
CC version: cc (GCC) 11.1.0

getwd: /home/sami
vmodules: /home/sami/.vmodules
vroot: /home/sami/v
vexe: /home/sami/v/v
vexe mtime: 2021-10-21 21:52:15
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 8d1ba52.5b69593

Git version: git version 2.33.1
Git vroot status: weekly.2021.42.1-8-g5b695937
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 333c14de

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg /home/sami/testv.v`
```v
struct Abc {}
type Test = Abc | int | bool
fn test(a Test) Test {
	return match a {
		Abc { 20 }
		int {
			match a {
				1 { true }
				2 { false }
				else { -1 }
			}
		}
		bool { 1 }
	}
}

fn main() {
	println(test(1))
	println(test(2))
	println(test(3))
	println(test(true))
}

```

**What did you expect to see?**

success


**What did you see instead?**
```
==================
/tmp/v/testv.13871994419167025142.tmp.c:10640: error: cannot convert 'struct main__Test' to 'unsigned char'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -1557,7 +1557,7 @@ fn (mut g Gen) stmt(node ast.Stmt) {
 			// }
 			old_is_void_expr_stmt := g.is_void_expr_stmt
 			g.is_void_expr_stmt = !node.is_expr
-			if node.typ != ast.void_type && g.expected_cast_type != 0 {
+			if node.typ != ast.void_type && g.expected_cast_type != 0 && node.expr !is ast.MatchExpr {
 				g.expr_with_cast(node.expr, node.typ, g.expected_cast_type)
 			} else {
 				g.expr(node.expr)

--- a/vlib/v/tests/match_expr_nested_test.v
+++ b/vlib/v/tests/match_expr_nested_test.v
@@ -1,0 +1,35 @@
+struct Abc {}
+
+type Test = Abc | bool | int
+
+fn test(a Test) Test {
+	return match a {
+		Abc {
+			20
+		}
+		int {
+			match a {
+				1 { true }
+				2 { false }
+				else { -1 }
+			}
+		}
+		bool {
+			1
+		}
+	}
+}
+
+fn test_nested_match_expr() {
+	println(test(1))
+	assert test(1) == Test(true)
+
+	println(test(2))
+	assert test(2) == Test(false)
+
+	println(test(3))
+	assert test(3) == Test(-1)
+
+	println(test(true))
+	assert test(true) == Test(1)
+}

